src/ChaosBox/CLI.hs

```diff
@@ -265,15 +277,55 @@ runChaosBoxDirectly Opts {..} doRender = replicateM_ optRenderTimes $ do
       , gcMetadataString = optMetadataString
       }
 
-  void . renderWith surface . flip runReaderT ctx . flip runRandT stdGen $ do
-    cairo $ scale optScale optScale
-    void doRender
+  if optFileFormat == PNG || optRenderMode == Interactive then
+    void . renderWith surface . flip runReaderT ctx . flip runRandT stdGen $ do
+      cairo $ scale optScale optScale
+      void doRender
+      
+      ref   <- asks gcBeforeSaveHook
+      mHook <- liftIO $ readIORef ref
+      fromMaybe (pure ()) mHook
 
-    ref   <- asks gcBeforeSaveHook
-    mHook <- liftIO $ readIORef ref
-    fromMaybe (pure ()) mHook
+      saveImage
 
-    saveImage
+  else case optFileFormat of
+    SVG -> write ctx stdGen (fromIntegral w) (fromIntegral h) withSVGSurface
+    PS  -> write ctx stdGen (fromIntegral w) (fromIntegral h) withPSSurface
+    PDF -> write ctx stdGen (fromIntegral w) (fromIntegral h) withPDFSurface
+    PNG -> error "impossible"
+
+  where write ctx stdGen w h withSurface = void $ do
+          let GenerateCtx {..} = ctx
+          let dotExtension = '.' : map toLower (show optFileFormat)
+
+          putStrLn "Saving Image"
+          createDirectoryIfMissing True $ "./images/" <> gcName
+
+          putStrLn "Generating art..."
+          let regularFile =
+                "images/"
+                  <> gcName
+                  <> "/"
+                  <> show gcSeed
+                  <> "-"
+                  <> show gcScale
+                  <> fromMaybe "" gcMetadataString
+                  <> dotExtension
+              latest = "images/" <> gcName <> "/latest" <> dotExtension
+
+          withSurface regularFile w h $ \surface -> do
+            putStrLn $ "Writing " <> regularFile
+            _ <- renderWith surface . flip runReaderT ctx . flip runRandT stdGen $ do
+              cairo $ scale optScale optScale
+              doRender
+            surfaceFinish surface
+            
+          withSurface latest w h $ \surface -> do
+            putStrLn $ "Writing " <> latest
+            _ <- renderWith surface . flip runReaderT ctx . flip runRandT stdGen $ do
+              cairo $ scale optScale optScale
+              doRender
+            surfaceFinish surface
```

Comment:
This will duplicate rendering which will often be very expensive; it would be better to write to `latest` and `cp` to the regular filepath.

Can you repurpose [saveImageWith](https://github.com/5outh/chaosbox/pull/14/files#diff-3ca7e6a79c5b77de86ce593f80780a15R335) to handle these new cases? It looks like you could replace your `write` above by changing the signature to:


```haskell
saveImageWith :: Surface -> Maybe FilePath -> IO ()
```

Reply:
I completely agree with the double rendering, I changed it such that it copies the file now! Regarding repurposing the `saveImageWith` function, it's not that easy :(

I tried it since I forgot why it didn't work last time, but in the end the biggest problem is that you need the file name for the `withSurface` function. I have been able to get rid of the redundancy in another way though. It's not super clean but I could not think of anything easier. If you can think of a better way let me know!